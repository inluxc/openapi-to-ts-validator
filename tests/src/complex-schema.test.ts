import path from "node:path";
import fs from "node:fs";
import { generate } from "openapi-to-ts-validator";
import Ajv from "ajv";

describe("complex-schema", () => {
  const name = "complex";
  const generatedDir = path.join(__dirname, "../generated", name);
  const schemaDir = path.join(__dirname, "../schemas");

  beforeAll(async () => {
    if (fs.existsSync(generatedDir))
      fs.rmSync(generatedDir, { recursive: true });
    await generate({
      schemaFile: path.join(schemaDir, "complex-schema.json"),
      schemaType: "json",
      directory: generatedDir,
      standalone: {
        validatorOutput: "module",
      },
    });
  });

  test("schema should match", async () => {
    const file = fs.readFileSync(
      path.join(generatedDir, "schema.json"),
      "utf8"
    );
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
    expect(await new Ajv().validateSchema(JSON.parse(file))).toEqual(true);
  });

  describe("decoders", () => {
    const decodersDir = path.join(generatedDir, "decoders");

    test("file structure", () => {
      const dir = fs.readdirSync(decodersDir);
      expect(dir).toEqual([
        "Component",
        "ImageComponent",
        "Screen",
        "TitleComponent",
        "index.ts",
      ]);
    });

    test("index.ts", () => {
      const file = fs.readFileSync(path.join(decodersDir, "index.ts"), "utf8");
      expect(file).toMatchSnapshot();
    });
  });

  test("models should match", () => {
    const file = fs.readFileSync(path.join(generatedDir, "models.ts"), "utf8");
    expect(file).not.toBeUndefined();
    expect(file).toMatchSnapshot();
  });
});
