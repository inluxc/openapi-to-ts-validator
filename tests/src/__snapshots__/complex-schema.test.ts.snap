// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex-schema decoders index.ts 1`] = `
"/* eslint-disable */

export { ScreenDecoder } from \"./Screen/decoder\";
export { ComponentDecoder } from \"./Component/decoder\";
export { TitleComponentDecoder } from \"./TitleComponent/decoder\";
export { ImageComponentDecoder } from \"./ImageComponent/decoder\";
"
`;

exports[`complex-schema models should match 1`] = `
"/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export type Component = TitleComponent | ImageComponent;

export interface Screen {
  components: Component[];
}
export interface TitleComponent {
  type: \"title\";
  title: string;
  subtitle?: string | null;
}
export interface ImageComponent {
  type: \"image\";
  url: string;
}
"
`;

exports[`complex-schema schema should match 1`] = `
"{
  \"type\": \"object\",
  \"title\": \"Schema\",
  \"definitions\": {
    \"Screen\": {
      \"type\": \"object\",
      \"properties\": {
        \"components\": {
          \"type\": \"array\",
          \"items\": {
            \"$ref\": \"#/definitions/Component\"
          }
        }
      },
      \"required\": [
        \"components\"
      ],
      \"$schema\": \"http://json-schema.org/draft-04/schema#\"
    },
    \"Component\": {
      \"anyOf\": [
        {
          \"$ref\": \"#/definitions/TitleComponent\"
        },
        {
          \"$ref\": \"#/definitions/ImageComponent\"
        }
      ],
      \"$schema\": \"http://json-schema.org/draft-04/schema#\"
    },
    \"TitleComponent\": {
      \"type\": \"object\",
      \"properties\": {
        \"type\": {
          \"type\": \"string\",
          \"enum\": [
            \"title\"
          ]
        },
        \"title\": {
          \"type\": \"string\"
        },
        \"subtitle\": {
          \"type\": [
            \"string\",
            \"null\"
          ]
        }
      },
      \"required\": [
        \"type\",
        \"title\"
      ],
      \"$schema\": \"http://json-schema.org/draft-04/schema#\"
    },
    \"ImageComponent\": {
      \"type\": \"object\",
      \"properties\": {
        \"type\": {
          \"type\": \"string\",
          \"enum\": [
            \"image\"
          ]
        },
        \"url\": {
          \"type\": \"string\"
        }
      },
      \"required\": [
        \"type\",
        \"url\"
      ],
      \"$schema\": \"http://json-schema.org/draft-04/schema#\"
    }
  },
  \"properties\": {
    \"Screen\": {
      \"$ref\": \"#/definitions/Screen\"
    },
    \"Component\": {
      \"$ref\": \"#/definitions/Component\"
    },
    \"TitleComponent\": {
      \"$ref\": \"#/definitions/TitleComponent\"
    },
    \"ImageComponent\": {
      \"$ref\": \"#/definitions/ImageComponent\"
    }
  }
}"
`;
